File: support/prototype/reader.go
package sheeter

import (
	"encoding/json"
	"fmt"
)

// Handmade $結構說明
type Handmade struct {
	Pkey   int32     `json:"Pkey"`   // $欄位說明
	Skey   string    `json:"Skey"`   // $欄位說明
	Data1  bool      `json:"Data1"`  // $欄位說明
	Data2  []bool    `json:"Data2"`  // $欄位說明
	Data3  int32     `json:"Data3"`  // $欄位說明
	Data4  []int32   `json:"Data4"`  // $欄位說明
	Data5  int64     `json:"Data5"`  // $欄位說明
	Data6  []int64   `json:"Data6"`  // $欄位說明
	Data7  float32   `json:"Data7"`  // $欄位說明
	Data8  []float32 `json:"Data8"`  // $欄位說明
	Data9  float64   `json:"Data9"`  // $欄位說明
	Data10 []float64 `json:"Data10"` // $欄位說明
	Data11 string    `json:"Data11"` // $欄位說明
	Data12 []string  `json:"Data12"` // $欄位說明
}

// HandmadeReader $結構說明
type HandmadeReader struct {
	Data map[string]*Handmade // $欄位說明
}

// FileName 取得檔名物件
func (this *HandmadeReader) FileName() FileName {
	return NewFileName("handmade", ".json")
}

// FromData 讀取資料
func (this *HandmadeReader) FromData(data []byte, clear bool, progress *Progress) error {
	tmpl := map[string]*Handmade{}

	if err := json.Unmarshal(data, &tmpl); err != nil {
		return fmt.Errorf("from data: %w", err)
	} // if

	if clear || this.Data == nil {
		this.Data = map[string]*Handmade{}
	} // if

	task := progress.Reg()
	curr := 0
	total := len(tmpl)

	for k, v := range tmpl {
		if _, ok := this.Data[k]; ok {
			return fmt.Errorf("from data: key duplicate [handmade : %v]", k)
		} // if

		this.Data[k] = v
		curr++
		progress.Set(task, curr, total)
	} // for

	return nil
}

// Clear 清除資料
func (this *HandmadeReader) Clear() {
	this.Data = map[string]*Handmade{}
}

// Get 取得資料
func (this *HandmadeReader) Get(key string) *Handmade {
	if result, ok := this.Data[key]; ok {
		return result
	} // if

	return nil
}

// Keys 取得索引列表
func (this *HandmadeReader) Keys() (result []string) {
	for itor := range this.Data {
		result = append(result, itor)
	} // for

	return result
}

// Values 取得資料列表
func (this *HandmadeReader) Values() (result []*Handmade) {
	for _, itor := range this.Data {
		result = append(result, itor)
	} // for

	return result
}

// ValuesAny 取得資料列表
func (this *HandmadeReader) ValuesAny() (result []any) {
	for _, itor := range this.Data {
		result = append(result, itor)
	} // for

	return result
}

// Count 取得資料數量
func (this *HandmadeReader) Count() int {
	return len(this.Data)
}

File: support/prototype/reader.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;

namespace Sheeter
{
    using Data_ = Handmade;
    using Key_ = String;
    using Store_ = Dictionary<String, Handmade>;

    /// <summary>
    /// $結構說明
    /// </summary>
    public partial class Handmade
    {
        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Pkey")]
        public Int32 Pkey { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Skey")]
        public String Skey { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data1")]
        public Boolean Data1 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data2")]
        public Boolean[] Data2 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data3")]
        public Int32 Data3 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data4")]
        public Int32[] Data4 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data5")]
        public Int64 Data5 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data6")]
        public Int64[] Data6 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data7")]
        public Single Data7 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data8")]
        public Single[] Data8 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data9")]
        public Double Data9 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data10")]
        public Double[] Data10 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data11")]
        public String Data11 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data12")]
        public String[] Data12 { get; set; }
    }

    /// <summary>
    /// $結構說明
    /// </summary>
    public partial class HandmadeReader : Reader
    {
        /// <summary>
        /// 取得檔名物件
        /// </summary>
        public FileName FileName()
        {
            return new FileName("handmade", ".json");
        }

        /// <summary>
        /// 讀取資料
        /// </summary>
        public string FromData(string data, bool clear, Progress progress)
        {
            Store_ tmpl;

            try
            {
                tmpl = JsonConvert.DeserializeObject<Store_>(data);
            } // try
            catch
            {
                return "from data: deserialize failed";
            } // catch

            if (tmpl == null)
                return "from data: deserialize failed";

            if (clear)
                this.data = new();

            var task = progress.Reg();
            var curr = 0;
            var total = tmpl.Count;

            foreach (var itor in tmpl)
            {
                if (this.data.ContainsKey(itor.Key))
                    return "from data: key duplicate [handmade : " + itor.Key + "]";

                this.data[itor.Key] = itor.Value;
                curr++;
                progress.Set(task, curr, total);
            } // for

            return string.Empty;
        }

        /// <summary>
        /// 清除資料
        /// </summary>
        public void Clear()
        {
            data.Clear();
        }

        /// <summary>
        /// 取得資料
        /// </summary>
        public bool TryGetValue(Key_ key, out Data_ value)
        {
            return data.TryGetValue(key, out value);
        }

        /// <summary>
        /// 檢查索引是否存在
        /// </summary>
        public bool ContainsKey(Key_ key)
        {
            return data.ContainsKey(key);
        }

        /// <summary>
        /// 取得迭代器
        /// </summary>
        public IEnumerator<KeyValuePair<Key_, Data_>> GetEnumerator()
        {
            return data.GetEnumerator();
        }

        /// <summary>
        /// 取得資料
        /// </summary>
        public Data_ this[Key_ key]
        {
            get { return data[key]; }
        }

        /// <summary>
        /// 取得索引列表
        /// </summary>
        public ICollection<Key_> Keys
        {
            get { return data.Keys; }
        }

        /// <summary>
        /// 取得資料列表
        /// </summary>
        public ICollection<Data_> Values
        {
            get { return data.Values; }
        }

        /// <summary>
        /// 取得資料數量
        /// </summary>
        public int Count
        {
            get { return data.Count; }
        }

        private Store_ data = new();
    }
}
